### PR TITLE
Refactor: simplify UI for data source selection (vault and assistant)

### PR DESCRIPTION
**Pull request by Fraggle**
## Description

Rework the assistant builder UI to be able to browse any datasourceview from a single screen.
- Re-used work from @PopDaph (renamed `VaultManagedDataSourceViewsTree` to `DataSourceViewSelector`)
- Use the same type `DataSourceViewSelectionConfigurations` (and `DataSourceViewSelectionConfiguration`) everywhere.
- Added an initial fetch to get `LightContentNode` from `DataSourceView`'s  `parentsIn` in Vault Datasource configuration.

## Risk

Assistant builder and Vault datasource selection screen could be broken.

## Deploy Plan

Deploy `Front-edge` then `Front`

### CODE REVIEW

File: front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx [hunk 1]
```diff
@@ -211,98 +40,39 @@ export default function AssistantBuilderDataSourceModal({
...
-            toggleSelectAll={onToggleSelectAll}
-          />
+        {orderDatasourceViewByImportance(dataSourceViews).map(
+          (dataSourceView) => (
+            <DataSourceViewSelector
+              key={dataSourceView.sId}
+              owner={owner}
+              selectionConfiguration={
+                selectionConfigurations[dataSourceView.sId] ??
+                defaultSelectionConfiguration(dataSourceView)
```
Comment (flvndvd) [inline front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx:66]: Not sure about the return value of this function, but I assume it returns a "new" value on every call? If so, we might want to memoize.

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx [hunk 2]
```diff
@@ -47,69 +47,109 @@ export function EditVaultManagedDataSourcesViews({
...
+            `/api/w/${owner.sId}/vaults/${vault.sId}/data_source_views/${deletedView.sId}`,
+            {
+              method: "DELETE",
+              headers: {
+                "Content-Type": "application/json",
+              },
+            }
+          );
+          if (!res.ok) {
+            const rawError = (await res.json()) as { error: APIError };
```
Comment (flvndvd) [inline front/components/vaults/EditVaultManagedDatasourcesViews.tsx:74]:
```suggestion
            const rawError: {error: APIError} = await res.json();
```

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx [hunk 3]
```diff
@@ -47,69 +47,109 @@ export function EditVaultManagedDataSourcesViews({
...
+            return rawError.error.message;
+          }
+        } catch (e) {
+          return null;
```
Comment (flvndvd) [inline front/components/vaults/EditVaultManagedDatasourcesViews.tsx:78]: Shouldn't we display an error if it fails?

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx [hunk 4]
```diff
@@ -47,69 +47,109 @@ export function EditVaultManagedDataSourcesViews({
...
+
+    const upsertPromisesErors = await Promise.all(
+      Object.values(selectionConfigurations).map(
+        async (selectionConfiguration) => {
+          const sDsv = selectionConfiguration.dataSourceView;
+          const sDs = sDsv.dataSource;
+          const existingViewForDs = vaultDataSourceViews.find(
+            (d) =>
+              d.dataSource.name ===
+              selectionConfiguration.dataSourceView.dataSource.name
```
Comment (flvndvd) [inline front/components/vaults/EditVaultManagedDatasourcesViews.tsx:91]:
```suggestion
              sDs.name
```

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx [hunk 5]
```diff
@@ -47,69 +47,109 @@ export function EditVaultManagedDataSourcesViews({
...
+        } catch (e) {
+          return null;
+        }
+      })
+    );
+
+    const upsertPromisesErors = await Promise.all(
+      Object.values(selectionConfigurations).map(
+        async (selectionConfiguration) => {
+          const sDsv = selectionConfiguration.dataSourceView;
```
Comment (flvndvd) [inline front/components/vaults/EditVaultManagedDatasourcesViews.tsx:86]: Here destructuring would help ease readability of the code.

File: front/components/vaults/VaultManagedDatasourcesViewsModal.tsx [hunk 6]
```diff
@@ -1,71 +1,119 @@
...
+                  key={dataSourceView.sId}
                   owner={owner}
-                  dataSourceView={dataSourceView}
-                  selectedNodes={selectedNodes}
-                  setSelectedNodes={setSelectedNodes}
-                  hasChanged={hasChanged}
-                  setHasChanged={setHasChanged}
+                  selectionConfiguration={
+                    selectionConfigurations[dataSourceView.sId] ??
+                    defaultSelectionConfiguration(dataSourceView)
```
Comment (flvndvd) [inline front/components/vaults/VaultManagedDatasourcesViewsModal.tsx:109]: Same here as my previous comment on `defaultSelectionConfiguration`.

File: front/lib/assistant.ts [hunk 7]
```diff
@@ -171,6 +171,14 @@ function compareByImportance(
           .connectorProvider
     );
 
+  if (aConnector === null && bConnector != null) {
```
Comment (flvndvd) [inline front/lib/assistant.ts:174]: This function becomes pretty hard to parse.

Can we:
1. extract the `order` compute since it does not rely on the function's parameters.
2 then we can probably do something like this:


```
function compareByImportance(
  a: ComparableByProvider,
  b: ComparableByProvider
): number {
  // Handle null cases.
  if (aConnector === null) return bConnector === null ? 0 : 1;
  if (bConnector === null) return -1;

  // Handle webcrawler cases.
  if (aConnector === webcrawlerProvider) return 1;
  if (bConnector === webcrawlerProvider) return -1;

  // Get indices in sorted connectors.
  const indexA = order.indexOf(aConnector);
  const indexB = order.indexOf(bConnector);

  // If both are not found, they are considered equal.
  if (indexA === -1 && indexB === -1) return 0;

  // Compare indices, treating not found as less important.
  return (indexA === -1 ? order.length : indexA) - (indexB === -1 ? order.length : indexB);
}
```

File: front/lib/swr.ts [hunk 8]
```diff
@@ -120,6 +120,17 @@ export const postFetcher = async ([url, body]: [string, object]) => {
   return resHandler(res);
 };
 
+type UrlsAndOptions = { url: string; options: object };
+
+const fetcherMultiple = <T>(urlsAndOptions: UrlsAndOptions[]) => {
+  const f = (url: string, options: object) =>
+    fetcher(url, options).then((r) => r);
```
Comment (flvndvd) [inline front/lib/swr.ts:127]: No thenable please.

File: front/lib/swr.ts [hunk 9]
```diff
@@ -120,6 +120,17 @@ export const postFetcher = async ([url, body]: [string, object]) => {
...
+
+  return Promise.all<T>(
+    urlsAndOptions.map(({ url, options }) => f(url, options))
```
Comment (flvndvd) [inline front/lib/swr.ts:130]: OOC what logic is then being used by the caching system of SWR?

File: front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx (hunk 1 above)
Comment (Fraggle) [reply front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx:66]: i don't think it's worth memoizing as it will be added to the `selectionConfigurations` and next time you'll not call it.

File: front/lib/swr.ts [hunk 10]
```diff
@@ -481,6 +492,64 @@ export function useDataSourceContentNodes({
...
+    fetcherMultiple<GetContentNodeResponseBody>
+  );
+
+  const isNodesError = errors?.some((e: boolean | undefined) => !!e) ?? false;
+  const isNodesLoading = results?.some((r) => !r.nodes) ?? true;
+
+  return useMemo(
+    () => ({
+      dataSourceViewsAndNodes: dataSourceViewsAndInternalIds.map(
+        ({ dataSourceView }, i) => ({
```
[the remark this answers is not in the document]
Comment (Fraggle) [reply front/lib/swr.ts:541]: I don't think so because if you need any other property of DataSourceViewType, you'll need to do some kind of mapping somewhere (eg: `.dataSource`)

File: front/lib/swr.ts (hunk 9 above)
Comment (Fraggle) [reply front/lib/swr.ts:130]: I don't know but it seems to work fine as the request is never retriggered.

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx [hunk 11]
```diff
@@ -71,24 +71,25 @@ export function EditVaultManagedDataSourcesViews({
             }
           );
           if (!res.ok) {
-            const rawError = (await res.json()) as { error: APIError };
+            const rawError: { error: APIError } = await res.json();
             return rawError.error.message;
           }
         } catch (e) {
-          return null;
```
Comment (flvndvd) [inline front/components/vaults/EditVaultManagedDatasourcesViews.tsx]: Seems a bit hacky no?

**Review by flvndvd — COMMENTED** [review]
LGTM, will stamped once I test on front-edge.

File: front/components/vaults/EditVaultManagedDatasourcesViews.tsx (hunk 11 above)
Comment (Fraggle) [reply front/components/vaults/EditVaultManagedDatasourcesViews.tsx]: I am open to suggestions here :)

**Review by flvndvd — COMMENTED** [review]
Seems like we don't display the right name for websites:
image

**Comment by Fraggle** [comment]
> Seems like we don't display the right name for websites:

Good catch, will fix.

**Review by flvndvd — APPROVED** [review]
